File: rigid_transform_3D.py
#!/usr/bin/env python3

import numpy as np


def rigid_transform_3D(src_pts, dst_pts):
    """Calculates the optimal rigid transform from src_pts to dst_pts.

    Parameters
    ------
    src_pts: 3xN or Nx3 matrix
    dst_pts: 3xN or Nx3 matrix

    NOTE: If src_pts and dst_pts are 3x3 matrices then points are assumed to be row major.

    Returns
    -------
    R = 3x3 rotation matrix
    t = 3x1 column vector
    """

    assert (
        src_pts.shape == dst_pts.shape
    ), f"src and dst points aren't the same shape {src_pts.shape=} {dst_pts.shape=}"
    assert (
        src_pts.shape[0] == 3 or dst_pts.shape[1] == 3
    ), "Expect 3xN or Nx3 matrix of points"

    # transpose to row major
    if src_pts.shape[0] == 3:
        src_pts = src_pts.T
        dst_pts = dst_pts.T

    # find mean/centroid
    centroid_src = np.mean(src_pts, axis=0)
    centroid_dst = np.mean(dst_pts, axis=0)

    centroid_src = centroid_src.reshape(-1, 3)
    centroid_dst = centroid_dst.reshape(-1, 3)

    # subtract mean
    # NOTE: doing src_pts -= centroid_src will modifiy input!
    src_pts = src_pts - centroid_src
    dst_pts = dst_pts - centroid_dst

    # this is similar to the cross-covariance matrix, except the outer mean is not calculated
    # https://en.wikipedia.org/wiki/Cross-covariance_matrix
    H = src_pts.T @ dst_pts

    assert H.shape[0] == 3 and H.shape[1] == 3, "H is not a 3x3 matrix"

    # sanity check
    rank = np.linalg.matrix_rank(H)
    if rank < 3:
        print(f"WARNING: rank of H = {rank}, expecting rank of 3 for a unique solution")

    # find rotation
    U, _, Vt = np.linalg.svd(H)
    R = Vt.T @ U.T

    # special reflection case
    # https://en.wikipedia.org/wiki/Kabsch_algorithm
    if np.linalg.det(R) < 0:
        print("det(R) < 1, reflection detected!, correcting for it ...")
        S = np.diag([1, 1, -1])
        R = Vt.T @ S @ U.T

    t = -R @ centroid_src.T + centroid_dst.T

    return R, t

File: rigid_transform_3D.m

function [R,t] = rigid_transform_3D(A, B)
    
    % This function finds the optimal Rigid/Euclidean transform in 3D space
    % It expects as input a 3xN matrix of 3D points.
    % It returns R, t

    narginchk(2,2);
    assert(all(size(A) == size(B)));

    [num_rows, num_cols] = size(A);
    if num_rows ~= 3
        error("matrix A is not 3xN, it is %dx%d", num_rows, num_cols)
    end

    [num_rows, num_cols] = size(B);
    if num_rows ~= 3
        error("matrix B is not 3xN, it is %dx%d", num_rows, num_cols)
    end

    % find mean column wise
    centroid_A = mean(A, 2);
    centroid_B = mean(B, 2);

    % subtract mean
    Am = A - repmat(centroid_A, 1, num_cols);
    Bm = B - repmat(centroid_B, 1, num_cols);

    % calculate covariance matrix (is this the correct terminology?)
    H = Am * Bm';

    %if rank(H) < 3
    %    error(sprintf("rank of H = %d, expecting 3", rank(H)))
    %end

    % find rotation
    [U,S,V] = svd(H);
    R = V*U';

    if det(R) < 0
        printf("det(R) < R, reflection detected!, correcting for it ...\n");
        V(:,3) = V(:,3) * -1;
        R = V*U';
    end

    t = -R*centroid_A + centroid_B;
end
